File: 622-design-circular-queue/622-design-circular-queue.cpp
class MyCircularQueue {
public:
    MyCircularQueue(int k) {
        a_size = k;
        cq.resize(k);
    }
    
    bool enQueue(int value) {
        if (isFull()) 
            return false;
        if (isEmpty()){
            head = tail = 0;
            cq[tail] = value;
            size++;
            return true;
        }
        tail++;
        tail %= a_size;
        cq[tail] = value;
        size++;
        return true;
    }
    
    bool deQueue() {
        if (isEmpty()) return false;
        
        head = (head + 1) % a_size;
        size--;
        return true;
    }
    
    int Front() {
        return isEmpty() ? -1 : cq[head];
    }
    
    int Rear() {
        return isEmpty() ? -1 : cq[tail];
    }
    
    bool isEmpty() {
        return size == 0;
    }
    
    bool isFull() {
        return size == a_size;
    }
    
    private:
        int head = 0, tail = 0, size = 0, a_size = 0;
        vector<int> cq;
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */

File: 622-design-circular-queue/622-design-circular-queue.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.q = [0] * k
        self.count = 0
        self.size = k
        self.head_Index = 0
        

    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False
        self.q[(self.head_Index + self.count) % self.size] = value
        self.count += 1
        return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False
        self.head_Index = (self.head_Index + 1) % self.size
        self.count -= 1
        return True

    def Front(self) -> int:
        if self.isEmpty():
            return -1
        return self.q[self.head_Index]

    def Rear(self) -> int:
        if self.isEmpty():
            return -1
        return self.q[(self.head_Index + self.count - 1) % self.size] 
        

    def isEmpty(self) -> bool:
        return self.count == 0

    def isFull(self) -> bool:
        return self.count == self.size


# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()